src/hotspot/share/compiler/compilerDefinitions.hpp

```diff
@@ -149,8 +133,120 @@ class CompilerConfig : public AllStatic {
+  }
+
+  static bool is_jvmci_compiler_available() {
+    return is_jvmci_compiler() && !is_interpreter_only() && !is_c1_only();
+  }
+  // Is the JVM in a configuration that permits only c2-compiled methods?
+  // JVMCI compiler replaces C2.
+  static bool is_c2_only() {
+    if (is_c2_available()) {
+      const bool c2_only = !has_c1();
+      // There is no JVMCI compiler to replace C2 in the broker, and the user (or ergonomics)
+      // is forcing C1 off.
+      const bool c2_only_compilation_mode = CompilationModeFlag::high_only();
+      const bool tiered_off = !TieredCompilation;
+      return c2_only || c2_only_compilation_mode || tiered_off;
+    }
+    return false;
+  }
+
+  // Is the JVM in a configuration that permits only jvmci-compiled methods?
+  static bool is_jvmci_compiler_only() {
+    if (is_jvmci_compiler_available()) {
+      const bool jvmci_compiler_only = !has_c1();
+      // JVMCI compiler replaced C2 and the user (or ergonomics) is forcing C1 off.
+      const bool jvmci_only_compilation_mode = CompilationModeFlag::high_only();
+      const bool tiered_off = !TieredCompilation;
+      return jvmci_compiler_only || jvmci_only_compilation_mode || tiered_off;
+    }
+    return false;
+  }
+
+  static bool is_c2_or_jvmci_compiler_only() {
+    return is_c2_only() || is_jvmci_compiler_only();
+  }
+
+  // Tiered is basically C1 & (C2 | JVMCI) minus all the odd cases with restrictions.
+  static bool is_tiered() {
+    assert(is_c1_simple_only() && is_c1_only() || !is_c1_simple_only(), "c1 simple mode must imply c1-only mode");
+    return has_tiered() && !is_interpreter_only() && !is_c1_only() && !is_c2_or_jvmci_compiler_only();
+  }
+
+  static bool is_c1_available() {
+    return has_c1() && !is_interpreter_only() && !is_c2_or_jvmci_compiler_only();
+  }
+
+  static bool is_c1_profiling() {
+    const bool c1_only_profiling = is_c1_only() && !is_c1_simple_only();
+    const bool tiered = is_tiered();
+    return c1_only_profiling || tiered;
+  }
+
+
+  static bool is_c2_or_jvmci_compiler_available() {
+    return is_c2_available() || is_jvmci_compiler_available();
+  }
+
+  static bool is_aot()               { return AOT_ONLY(has_aot() && UseAOT) NOT_AOT(false);                 }
+  static bool is_jvmci_compiler()    { return JVMCI_ONLY(has_jvmci() && UseJVMCICompiler) NOT_JVMCI(false); }
+  static bool is_jvmci()             { return JVMCI_ONLY(has_jvmci() && EnableJVMCI) NOT_JVMCI(false);      }
+  static bool is_interpreter_only() {
+    return Arguments::is_interpreter_only() || TieredStopAtLevel == CompLevel_none;
+  }
+
+  static bool is_c1_or_interpreter_only_no_aot_or_jvmci() {
+    assert(is_jvmci_compiler() && is_jvmci() || !is_jvmci_compiler(), "JVMCI compiler implies enabled JVMCI");
+    return !is_aot() && !is_jvmci() && (is_interpreter_only() || is_c1_only());
+  }
+
+  static bool is_c1_only_no_aot_or_jvmci() {
+    return is_c1_only() && !is_aot() && !is_jvmci();
+  }
```

Comment:
I guess it's a matter of naming convention. What I tried to make the CompilerConfig API about is to check if compilers are present/enabled and in which combination. Of course presence of a compiler implies that we're going to see nmethod produced by it. I'd like to keep the current naming if it's not a huge eyesore for you.

Reply:
Hmm, may be we don't need these funxctions.
Seems is_c1_only() will be true only when JVMCI is off. Right? We can also reinforce it to exclude AOT too.
And AOT and JVMCI are disabled with -Xint.

Reply:
It's not so easy. You can have AOT or JVMCI enabled with -Xint. JVMCI code installs can be initiated not only as a result of method going through the compiler broker. Remember that you can run Graal in "host" mode, where it is a normal C1/C2 system (or may be just an interpreter); Truffle can work in a mode like that. And you can totally run with AOT and interpreter (without any JIT).

Reply:
I don't think so:
```
> java -Xint -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI -version
Java HotSpot(TM) 64-Bit Server VM warning: JVMCI Compiler disabled due to -Xint.

> java -Xint -XX:+UnlockExperimentalVMOptions -XX:+UseAOT -XX:+PrintAOT -version
Java HotSpot(TM) 64-Bit Server VM warning: -Xint is not compatible with AOT (switching AOT off)
```

Reply:
You're right. Perhaps I was overly defensive. It doesn't hurt though, less potential surprises in the future (as you correctly noticed those are used to make sure there are no nmethods produced by any other compiler but C1).
Also you still can have a C1 and AOT combination.

Reply:
Yes. You can have C1 in combination with AOT and JVMCI (hosted).
I looked on use cases for these 4 methods and it seems it will be hard to only use one.
Okay. Leave them as it is. May be add comments to explain why you needs these variants.